c/tskit/trees.c

```diff
@@ -4685,3 +4729,278 @@ tsk_tree_kc_distance(tsk_tree_t *self, tsk_tree_t *other, double lambda, double
+            for (c1 = t->left_child[v]; c1 != TSK_NULL; c1 = t->right_sib[c1]) {
+                stack_top++;
+                stack[stack_top].node = c1;
+                stack[stack_top].depth = depth + 1;
+
+                for (c2 = t->right_sib[c1]; c2 != TSK_NULL; c2 = t->right_sib[c2]) {
+                    update_kc_vectors_all_pairs(
+                        t, kc, c1, c2, depth, root_time - times[v]);
+                }
+            }
+        }
+    }
+
+out:
+    tsk_safe_free(stack);
+    return ret;
+}
+
+static int
+tsk_tree_node_root(tsk_tree_t *self, tsk_id_t u)
+{
+    tsk_id_t v = u;
+    while (self->parent[v] != TSK_NULL) {
+        v = self->parent[v];
+    }
+
+    return v;
+}
+
+/* TODO Should push this through the stack since this is a python method and
+ * it's here anyway */
+static double
+tsk_tree_branch_length(tsk_tree_t *self, tsk_id_t u)
+{
+    double *times = self->tree_sequence->tables->nodes.time;
+    tsk_id_t parent = self->parent[u];
+
+    return parent == TSK_NULL ? 0 : times[parent] - times[u];
+}
+
+static int
+update_kc_incremental(tsk_tree_t *self, kc_vectors *kc, tsk_edge_list_t *edges_out,
+    tsk_edge_list_t *edges_in)
+{
+    int ret = 0;
+    tsk_edge_list_t *e_list;
+    tsk_edge_t *e;
+    tsk_id_t u;
+    double root_time, time;
+    double *times = self->tree_sequence->tables->nodes.time;
+
+    for (e_list = edges_out; e_list != NULL; e_list = e_list->next) {
+        e = &e_list->edge;
+        u = e->child;
+        root_time = times[tsk_tree_node_root(self, u)];
+        ret = update_in_subtree_pairs(self, kc, u, root_time);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+    for (e_list = edges_in; e_list != NULL; e_list = e_list->next) {
+        e = &e_list->edge;
+        u = e->child;
+
+        if (tsk_tree_is_sample(self, u)) {
+            time = tsk_tree_branch_length(self, u);
+            update_kc_vectors_single_leaf(self->tree_sequence, kc, u, time);
+        }
+
+        root_time = times[tsk_tree_node_root(self, u)];
+        update_in_out_subtree_pairs(self, kc, e, root_time);
+        ret = update_in_subtree_pairs(self, kc, u, root_time);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+out:
+    return ret;
+}
+
+int
+tsk_treeseq_kc_distance(
+    tsk_treeseq_t *self, tsk_treeseq_t *other, double lambda_, double *result)
+{
+    int i;
+    tsk_id_t n;
+    double left, span, total;
+    tsk_tree_t t1, t2;
+    kc_vectors kc1, kc2;
+    tsk_diff_iter_t diff_iter1, diff_iter2;
+    tsk_treeseq_t *treeseqs[2] = { self, other };
+    tsk_tree_t *trees[2] = { &t1, &t2 };
+    kc_vectors *kcs[2] = { &kc1, &kc2 };
+    tsk_diff_iter_t *diff_iters[2] = { &diff_iter1, &diff_iter2 };
+    tsk_edge_list_t *t1_edges_out, *t1_edges_in, *t2_edges_out, *t2_edges_in;
+    double t1_left, t1_right, t2_left, t2_right;
+    int ret = 0;
+
+    for (i = 0; i < 2; i++) {
+        memset(trees[i], 0, sizeof(tsk_tree_t));
```

Comment:
Not convinced this is correct at the moment - isn't this writing ``sizeof(tsk_tree_t)`` bytes into a pointer sized variable?

Reply:
I _think_ this is correct? `trees` is an array of pointers and `memset` takes a pointer to the memory of the given size, so this should be passing a pointer to the memory of the `tsk_tree_t` struct. If `trees` were an array of structs it would be `&trees[i]` which I think is the same thing right?

Reply:
Either way I think I will make the change to just have an in-place array of structs because the end of the function doesn't get much messier with the change and it avoids this awkward duplication of pointers.

Reply:
Yes, you're right, it is correct. I think it's good to have the array of structs all right though, it's less confusing.